src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,372 @@
+                var symbolInfo = semanticModel.GetSymbolInfo(invocationNode, cancellationToken);
+                var candidateSymbols = symbolInfo.CandidateSymbols;
+
+                foreach (var candidcateSymbol in candidateSymbols)
+                {
+                    if (!(candidcateSymbol is IMethodSymbol methodSymbol))
+                    {
+                        continue;
+                    }
+
+                    // Test if all arguments can match all parameters, otherwise it is not the perfect match function. For example:
+                    // class Base { }
+                    // class Derived1 : Base { }
+                    // class Derived2 : Base { }
+                    // void DoSomething(int i, Derived1 d) { }
+                    // void DoSomething(string s, Derived2 d) { }
+                    // 
+                    // Base b;
+                    // DoSomething(1, [||]b);
+                    //
+                    // *void DoSomething(string s, Derived2 d) { }* is not the perfect match candidate function for
+                    // *DoSomething(1, [||]b)* because int and string are not ancestor-descendant relationship. Thus,
+                    // Derived2 is not a potential conversion type
+                    if (IsArgumentListAndParameterListPerfactMatch(semanticModel, argumentList.Arguments, methodSymbol.Parameters, targetArgument, cancellationToken, out var paramIndex))
+                    {
+                        var correspondingParameter = methodSymbol.Parameters[paramIndex];
+                        var argumentConversionType = correspondingParameter.Type;
+
+                        if (correspondingParameter.IsParams && correspondingParameter.Type is IArrayTypeSymbol arrayType && !(targetNodeType is IArrayTypeSymbol))
+                        {
+                            // target argument is matched to the parameter with keyword params
+                            argumentConversionType = arrayType.ElementType;
+                        }
+
+                        mutablePotentialConversionTypes.Add(argumentConversionType);
+                    }
+                }
+
+                // Sort the potential conversion types by inheritance distance
+                var comparer = new InheritanceDistanceComparer(semanticModel, targetNodeType);
+                mutablePotentialConversionTypes.Sort(comparer);
+                mutablePotentialConversionTypes = mutablePotentialConversionTypes.Distinct().ToList(); // clear up duplicate types
+            }
+
+            // For cases like object creation expression. for example:
+            // Derived d = [||]new Base();
+            // It is always invalid except the target node has explicit conversion operator.
+            // filter function is going to filter the cases that are invalid
+            var nodeType = targetNodeType;
+            Func<ITypeSymbol, bool> filter = (targetNodeConversionType =>
```

Comment:
1. remove outer parens.
2. using linq here just seems pretty convoluted.  i would jsut avoid it and iterate manually.